libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -187,37 +252,60 @@ void AP_Frsky_Telem::process_packet(uint8_t idx)
         case PARAM: // 0x5007 parameters
             send_sport_frame(SPORT_DATA_FRAME, DIY_FIRST_ID+7, calc_param());
             break;
+#if HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+        case MAV: // mavlite
+            process_sport_tx_queue();
+            break;
+#endif //HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL            
     }
 }
 
+/*
+ * true if we need to respond to the last polling byte
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+ */
+bool AP_Frsky_Telem::is_passthrough_byte()
+{
+#if HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+    return _passthrough.new_byte == SENSOR_ID_27 || _passthrough.new_byte == _SPort_bidir.downlink1_sensor_id || _passthrough.new_byte == _SPort_bidir.downlink2_sensor_id ;
+#else   //HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+    return _passthrough.new_byte == SENSOR_ID_27;
+#endif  //HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+}
+
+#if HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+bool AP_Frsky_Telem::parse_sport_rx_data() 
+{
+    return FRSky::parse_sport_telemetry_data(_passthrough.new_byte, _SPort_bidir.parse_state);
+}
+#endif //HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+
 /*
  * send telemetry data
  * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
  */
 void AP_Frsky_Telem::send_SPort_Passthrough(void)
 {
-    int16_t numc;
-    numc = _port->available();
-
-    // check if available is negative
-    if (numc < 0) {
-        return;
-    }
+    const uint16_t numc = MIN(_port->available(), 1024U);
 
     // this is the constant for hub data frame
     if (_port->txspace() < 19) {
         return;
     }
     // keep only the last two bytes of the data found in the serial buffer, as we shouldn't respond to old poll requests
     uint8_t prev_byte = 0;
-    for (int16_t i = 0; i < numc; i++) {
+    for (uint16_t i = 0; i < numc; i++) {
         prev_byte = _passthrough.new_byte;
         _passthrough.new_byte = _port->read();
-    }
-    if (prev_byte == FRAME_HEAD) {
-        if (_passthrough.new_byte == SENSOR_ID_27) { // byte 0x7E is the header of each poll request
-            run_wfq_scheduler();
+#if HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+        if (parse_sport_rx_data()) {
+            queue_sport_rx_packet();
         }
+#endif //HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+    }
+    // check if we should respond to this polling byte
+    if (prev_byte == FRAME_HEAD && is_passthrough_byte()) {
```

Comment:
This should  pass in _passthrough.new_byte  - or be named `new_byte_is_passthrough_byte`

Reply:
ok

> Can confirm the reboot issuegetting back UNSUPPORTED.
> 
> Airspeed/Baro seems to be intermittently failing; worked on a retry, first time did seem to get an error.

Reboot is no longer supported, it used to be when I used that terrible hack of a private gcs mavlink backend, once we find a nice way to access protected members of gcs_mavlink I can add rebbot back in.

As for the reply thing it just happens, the quality of transport is quite poor, especially for fport, from FC to TX I experience a constant 15% packet loss which means that a mavlite packet every 8-10 is corrupted, open to new ideas to handle all this